src/cpp/shared/adventure/adv.cpp

```diff
@@ -118,9 +121,53 @@ void game::MakeAllWaterVisible(int player) {
 }
 
 void advManager::DoEvent(class mapCell *cell, int locX, int locY) {
+  hero *hro = &gpGame->heroes[gpCurPlayer->curHeroIdx];
   int locType = cell->objType & 0x7F;
+  SAMPLE2 res2 = NULL_SAMPLE2;
   ScriptCallback("OnLocationVisit", locType, locX, locY);
-  this->DoEvent_orig(cell, locX, locY);
+
+  switch (locType) {
+    case LOCATION_SHRINE_FIRST: {
+      sprintf(gText, "{Shrine of the 1st Circle}\n\nYou come across a small shrine attended by a group of novice acolytes.  In exchange for your protection, they agree to teach you a simple spell - '%s'.  ", gSpellNames[cell->extraInfo - 1]);
+      break;
```

Comment:
This is spaghetti. You can factor these out into a separate "HandleSpellShrine" function to make the control-flow easier to follow.